fsharp-backend/src/LibBackend/StaticAssets.fs

```diff
@@ -107,8 +97,115 @@ let latestDeployHash (canvasID : CanvasID) : Task<Option<string>> =
   |> Sql.parameters [ "canvasID", Sql.uuid canvasID; "branch", Sql.string branch ]
   |> Sql.executeRowOptionAsync (fun read -> read.string "deploy_hash")
 
+let allDeploysInCanvas
+  (canvasName : CanvasName.T)
+  (canvasID : CanvasID)
+  : Task<List<StaticDeploy>> =
+  Sql.query
+    "SELECT deploy_hash, created_at, live_at
+     FROM static_asset_deploys
+     WHERE canvas_id=@canvasID
+     ORDER BY created_at
+     DESC LIMIT 25"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID ]
+  |> Sql.executeAsync (fun read ->
+    let deployHash = read.string "deploy_hash"
+
+    let status, lastUpdate =
+      match read.instantOrNone "live_at" with
+      | Some datetime -> Deployed, datetime
+      | None -> Deploying, read.instant "created_at"
+
+    { deployHash = deployHash
+      url = url canvasName deployHash Short
+      status = status
+      lastUpdate = lastUpdate })
+
+/// Creates a record of a new static asset deploy, returning a deployHash to be
+/// used by the actual upload process.
+let startStaticAssetDeploy
+  (user : Account.UserInfo)
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  : Task<string> =
+
+  // we include .fff (milliseconds) to ensure we don't encoutner conflicts,
+  // especially relevant to unit tests which record multiple deploys quickly.
+  let now = System.DateTime.Now.ToString("MM/dd/yyyy hh:mm:ss.fff tt")
+
+  let deployHash =
+    $"{canvasID}{now}"
+    |> sha1digest
+    |> Base64.urlEncodeToString
+    |> String.removeSuffix "="
+    |> String.toLowercase
+    |> String.take 10
+
+  Sql.query
+    "INSERT INTO static_asset_deploys
+      (canvas_id, branch, deploy_hash, uploaded_by_account_id)
+    VALUES (@canvasID, @branch, @deployHash, @uploadedBy)"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "branch", Sql.string branch
+                      "deployHash", Sql.string deployHash
+                      "uploadedBy", Sql.uuid user.id ]
+  |> Sql.executeNonQueryAsync
+  |> Task.map (fun _ -> deployHash)
+
+// CLEANUP: return an Error if the deploy hash doesn't exist
+// CLEANUP: decide what to do if the deploy is already finished
+let finishStaticAssetDeploy
+  (canvasID : CanvasID)
+  (canvasName : CanvasName.T)
+  (deployHash : string)
+  : Task<StaticDeploy> =
+  Sql.query
+    "UPDATE static_asset_deploys
+      SET live_at = NOW()
+      WHERE canvas_id = @canvasID AND deploy_hash = @deployHash
+      RETURNING live_at"
+  |> Sql.parameters [ "canvasID", Sql.uuid canvasID
+                      "deployHash", Sql.string deployHash ]
+  |> Sql.executeRowAsync (fun reader -> reader.instant "live_at")
+  |> Task.map (fun lastUpdate ->
```

Comment:
I'm unclear why we're creating this record here?

Reply:
I thought I'd handled this before - thanks for reminding me.